modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkByScrollAction.java

```diff
@@ -129,57 +128,44 @@ public void start() {
             return;
         }
         try {
-            // Default to sorting by _doc if it hasn't been changed.
-            if (firstSearchRequest.source().sorts() == null) {
-                firstSearchRequest.source().sort(fieldSort("_doc"));
-            }
             startTime.set(System.nanoTime());
-            if (logger.isDebugEnabled()) {
-                logger.debug("executing initial scroll against {}{}",
-                        firstSearchRequest.indices() == null || firstSearchRequest.indices().length == 0 ? "all indices"
-                                : firstSearchRequest.indices(),
-                        firstSearchRequest.types() == null || firstSearchRequest.types().length == 0 ? ""
-                                : firstSearchRequest.types());
-            }
+            scrollSource.start(response -> {
+                logger.debug("[{}] documents match query", response.getTotalHits());
+                onScrollResponse(timeValueSeconds(0), response);
+            });
         } catch (Throwable t) {
             finishHim(t);
             return;
         }
-        searchWithRetry(listener -> client.search(firstSearchRequest, listener), (SearchResponse response) -> {
-            logger.debug("[{}] documents match query", response.getHits().getTotalHits());
-            onScrollResponse(timeValueSeconds(0), response);
-        });
     }
 
     /**
      * Process a scroll response.
      * @param delay how long to delay processesing the response. This delay is how throttling is applied to the action.
-     * @param searchResponse the scroll response to process
+     * @param response the scroll response to process
      */
-    void onScrollResponse(TimeValue delay, SearchResponse searchResponse) {
+    void onScrollResponse(TimeValue delay, ScrollableHitSource.Response response) {
         if (task.isCancelled()) {
             finishHim(null);
             return;
         }
-        setScroll(searchResponse.getScrollId());
         if (    // If any of the shards failed that should abort the request.
-                (searchResponse.getShardFailures() != null && searchResponse.getShardFailures().length > 0)
+                (response.getFailures().size() > 0)
                 // Timeouts aren't shard failures but we still need to pass them back to the user.
-                || searchResponse.isTimedOut()
+                || response.isTimedOut()
                 ) {
-            startNormalTermination(emptyList(), unmodifiableList(Arrays.asList(searchResponse.getShardFailures())),
-                    searchResponse.isTimedOut());
+            startNormalTermination(emptyList(), response.getFailures(), response.isTimedOut());
```

Comment:
Sorry, not part of this PR but I find this method name misleading... maybe something like `refreshAndFinish()` ? or just `stop()` since we have a `start()` method?